File: solutions/java/HalloweenSale.java
import java.math.*;

public class HalloweenSale{
	public static int howManyGames(int p, int d, int m, int s) {
        if(s < p) return 0;
        
        double lastN = (((m - p)/(float)d * (-1)) + 1);
        
        if(lastN != Math.ceil(lastN)){
            lastN = Math.floor(lastN);
        }
        
        double an = p + (double)(lastN - 1) * (d * -1);
        
        double v
        
        double numberOfGames = 0;
        
        if(sum > s){
            numberOfGames = sum / s;
        }else{
            numberOfGames = ((p - m) / d) + 1 + ((s - sum) / m);
        }
        
        return (int)Math.floor(numberOfGames);
    }
}

File: solutions/kotlin/HalloweenSale.kt
fun howManyGames(p: Int, d: Int, m: Int, s: Int): Int {
    if(s < p) return 0 
    
    var lastN = (((m - p)/d.toDouble() * (-1)) + 1)
    
    if(lastN != Math.ceil(lastN)){
        lastN = Math.floor(lastN)
    }
    val an = p + (lastN - 1).toDouble() * (d * -1)
    
    val sum = ((p + an) * lastN) / 2.toDouble();
    
    val numberOfGames = if(sum > s){
        sum / s
    }else{
        ((p - m) / d) + 1 + ((s - sum) / m)
    }
    
    return Math.floor(numberOfGames).toInt()
}